ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java

```diff
@@ -164,307 +219,272 @@ private HelixBootstrapUpgradeUtil(String hardwareLayoutPath, String partitionLay
+   * Create an {@link InstanceConfig} for the given node from the static cluster information.
+   * @param node the {@link DataNodeId}
+   * @param partitionToInstances the map of partitions to instances that will be populated for this instance.
+   * @return the constructed {@link InstanceConfig}
    */
-  private boolean updateReplicaCapacityIfRequired(String partitionName, InstanceConfig instanceConfig, String mountPath,
-      long actualReplicaCapacity) {
-    boolean instanceConfigUpdated = false;
-    Map<String, String> diskInfo = instanceConfig.getRecord().getMapField(mountPath);
-    String currentReplicasStr = diskInfo.get(ClusterMapUtils.REPLICAS_STR);
-    StringBuilder newReplicaStrBuilder = new StringBuilder();
-    List<String> replicaInfoList = Arrays.asList(currentReplicasStr.split(ClusterMapUtils.REPLICAS_DELIM_STR));
-    for (String replicaInfo : replicaInfoList) {
-      String[] info = replicaInfo.split(ClusterMapUtils.REPLICAS_STR_SEPARATOR);
-      if (info[0].equals(partitionName)) {
-        long capacityInHelix = Long.valueOf(info[1]);
-        if (capacityInHelix != actualReplicaCapacity) {
-          info[1] = Long.toString(actualReplicaCapacity);
+  private InstanceConfig createInstanceConfigFromStaticInfo(DataNodeId node,
+      Map<String, Set<String>> partitionToInstances) {
+    String instanceName = getInstanceName(node);
+    InstanceConfig instanceConfig = new InstanceConfig(instanceName);
+    instanceConfig.setHostName(node.getHostname());
+    instanceConfig.setPort(Integer.toString(node.getPort()));
+    if (node.hasSSLPort()) {
+      instanceConfig.getRecord().setSimpleField(ClusterMapUtils.SSLPORT_STR, Integer.toString(node.getSSLPort()));
+    }
+    instanceConfig.getRecord().setSimpleField(ClusterMapUtils.DATACENTER_STR, node.getDatacenterName());
+    instanceConfig.getRecord().setSimpleField(ClusterMapUtils.RACKID_STR, node.getRackId());
+    instanceConfig.getRecord()
+        .setSimpleField(ClusterMapUtils.SCHEMA_VERSION_STR, Integer.toString(ClusterMapUtils.CURRENT_SCHEMA_VERSION));
+    // xid support has not yet been added.
+
+    List<String> sealedPartitionsList = new ArrayList<>();
+
+    Map<String, Map<String, String>> diskInfos = new HashMap<>();
+    for (HashMap.Entry<DiskId, SortedSet<Replica>> diskToReplicas : instanceToDiskReplicasMap.get(instanceName)
+        .entrySet()) {
+      DiskId disk = diskToReplicas.getKey();
+      SortedSet<Replica> replicasInDisk = diskToReplicas.getValue();
+      // Note: An instance config has to contain the information for each disk about the replicas it hosts.
+      // This information will be initialized to the empty string - but will be updated whenever the partition
+      // is added to the cluster.
+      String replicasStr = "";
+      for (ReplicaId replicaId : replicasInDisk) {
+        Replica replica = (Replica) replicaId;
+        replicasStr +=
```

Comment:
Maybe use `StringBuilder` here instead of String concatenation to reduce copies of string.

Reply:
Done.